File: 2685.Count the Number of Complete Components.java
class Solution {

    private List<Integer>[] graph;

    private boolean[] visited;

    private int vertices;

    private int degreeSum;

    public int countCompleteComponents(
            int numberOfNodes,
            int[][] edges
    ) {

        graph = new ArrayList[numberOfNodes];

        for (int i = 0; i < numberOfNodes; i++) {
            graph[i] = new ArrayList<>();
        }

        // Build adjacency list
        for (int[] edge : edges) {

            int nodeA = edge[0];
            int nodeB = edge[1];

            graph[nodeA].add(nodeB);
            graph[nodeB].add(nodeA);
        }

        visited = new boolean[numberOfNodes];

        int completeComponents = 0;

        for (int node = 0; node < numberOfNodes; node++) {

            if (visited[node]) {
                continue;
            }

            vertices = 0;
            degreeSum = 0;

            dfs(node);

            if (degreeSum == vertices * (vertices - 1)) {
                completeComponents++;
            }
        }

        return completeComponents;
    }

    private void dfs(int currentNode) {

        visited[currentNode] = true;

        vertices++;

        degreeSum += graph[currentNode].size();

        for (int neighbor : graph[currentNode]) {

            if (!visited[neighbor]) {
                dfs(neighbor);
            }
        }
    }
}

File: 2685. Count the Number of Complete Components.cpp
class Solution {
public:
    int countCompleteComponents(
        int numberOfNodes,
        vector<vector<int>>& edges
    ) {

        // Build adjacency list
        vector<vector<int>> graph(numberOfNodes);

        for (const auto& edge : edges) {

            int nodeA = edge[0];
            int nodeB = edge[1];

            graph[nodeA].push_back(nodeB);
            graph[nodeB].push_back(nodeA);
        }

        bitset<51> visited;

        int completeComponents = 0;

        for (int node = 0; node < numberOfNodes; node++) {

            if (visited.test(node)) {
                continue;
            }

            int vertices = 0;
            int degreeSum = 0;

            auto dfs = [&](auto&& dfs, int currentNode) -> void {

                visited.set(currentNode);

                vertices++;

                degreeSum += graph[currentNode].size();

                for (int neighbor : graph[currentNode]) {

                    if (!visited.test(neighbor)) {
                        dfs(dfs, neighbor);
                    }
                }
            };

            dfs(dfs, node);

            // A complete graph with V vertices
            // has V * (V - 1) total degree.
            if (degreeSum == vertices * (vertices - 1)) {
                completeComponents++;
            }
        }

        return completeComponents;
    }
};
